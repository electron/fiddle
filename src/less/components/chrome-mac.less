.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
