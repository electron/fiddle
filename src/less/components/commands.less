@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
