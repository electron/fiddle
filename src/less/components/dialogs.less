@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
