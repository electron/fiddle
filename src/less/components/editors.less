.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
