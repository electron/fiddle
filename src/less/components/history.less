.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
