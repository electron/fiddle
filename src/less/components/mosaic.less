@import '../variables.less';

.fiddle {
  .mosaic-window-toolbar {
    background: var(--background-1);
    box-shadow: inset 0 0 100px 100px rgba(53, 53, 53, 0.1);
    border-top: 1px solid rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-toolbar.draggable:hover {
    box-shadow: inset 0 0 100px 100px rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-title {
    color: var(--text-color-3);
  }

  button.mosaic-default-control {
    padding: 0;
    height: 20px;
    width: 20px;
    margin: 5px;
    background-color: #314246;
    border: 0.1rem solid #222b35;
  }

  button.mosaic-default-control:hover {
    background-color: #131d1f;
  }

  .mosaic-window-body {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: stretch;
    align-content: stretch;
    overflow: visible;
    z-index: 4;
    background-color: @background-4;
  }

  .mosaic.mosaic-blueprint-theme {
    background: #bdbdbd15;
  }

  .editorContainer,
  .monaco-editor {
    width: 100%;
    height: 100%;
  }

  .editorContainer {
    background: @background-1;
  }
}
