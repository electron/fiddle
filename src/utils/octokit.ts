import { Octokit } from '@octokit/rest';

import { AppState } from '../renderer/state';

let _octo: Octokit;

/**
 * Returns a loaded Octokit. If state is passed and authentication
 * is available, we'll token-authenticate.
 *
 * @export
 * @returns {Promise<typeof Octokit>}
 */
export async function getOctokit(appState?: AppState): Promise<Octokit> {
  // It's possible to load Gists without being authenticated,
  // but we get better rate limits when authenticated.
  _octo =
    _octo || appState?.gitHubToken
      ? new Octokit({
          auth: appState?.gitHubToken,
        })
      : new Octokit();

  return _octo;
}
