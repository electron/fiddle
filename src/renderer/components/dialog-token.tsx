import * as React from 'react';

import { Button, Callout, Dialog, InputGroup, Intent } from '@blueprintjs/core';
import { clipboard, shell } from 'electron';
import { observer } from 'mobx-react';

import { getOctokit } from '../../utils/octokit';
import { AppState } from '../state';

interface TokenDialogProps {
  appState: AppState;
}

interface TokenDialogState {
  tokenInput: string;
  verifying: boolean;
  error: boolean;
}

const TOKEN_SCOPES = ['gist'].join();
const TOKEN_DESCRIPTION = encodeURIComponent('Fiddle Gist Token');
const GENERATE_TOKEN_URL = `https://github.com/settings/tokens/new?scopes=${TOKEN_SCOPES}&description=${TOKEN_DESCRIPTION}`;

/**
 * The token dialog asks the user for a GitHub Personal Access Token.
 * It's also responsible for checking if the token is correct.
 *
 * @export
 * @class TokenDialog
 * @extends {React.Component<TokenDialogProps, TokenDialogState>}
 */
export const TokenDialog = observer(
  class TokenDialog extends React.Component<
    TokenDialogProps,
    TokenDialogState
  > {
    constructor(props: TokenDialogProps) {
      super(props);

      this.state = {
        verifying: false,
        error: false,
        tokenInput: '',
      };

      this.onSubmitToken = this.onSubmitToken.bind(this);
      this.openGenerateTokenExternal = this.openGenerateTokenExternal.bind(
        this,
      );
      this.onTokenInputFocused = this.onTokenInputFocused.bind(this);
      this.handleChange = this.handleChange.bind(this);
      this.onClose = this.onClose.bind(this);
    }

    /**
     * Handles the submission of a token
     *
     * @returns {Promise<void>}
     * @memberof TokenDialog
     */
    public async onSubmitToken(): Promise<void> {
      if (!this.state.tokenInput) return;
      this.setState({ verifying: true });
      this.props.appState.gitHubToken = this.state.tokenInput;

      const octo = await getOctokit(this.props.appState);

      try {
        const { data } = await octo.users.getAuthenticated();
        this.props.appState.gitHubAvatarUrl = data.avatar_url;
        this.props.appState.gitHubLogin = data.login;
        this.props.appState.gitHubName = data.name;
        this.setState({ verifying: false, error: false });
      } catch (error) {
        console.warn(`Authenticating against GitHub failed`, error);
        this.setState({ verifying: false, error: true });
        this.props.appState.gitHubToken = null;
        return;
      }

      this.props.appState.isTokenDialogShowing = false;
    }

    /**
     * Closes the dialog
     */
    public onClose() {
      this.props.appState.isTokenDialogShowing = false;
      this.reset();
    }

    /**
     * Reset this component's state
     */
    public reset(): void {
      this.setState({
        verifying: false,
        error: false,
        tokenInput: '',
      });
    }

    /**
     * Opens GitHub's page for token generation
     *
     * @memberof TokenDialog
     */
    public openGenerateTokenExternal() {
      shell.openExternal(GENERATE_TOKEN_URL);
    }

    /**
     * When the input field receives focus, we check the clipboard.
     * Maybe there's already something token-like there!
     *
     * @returns
     * @memberof TokenDialog
     */
    public onTokenInputFocused() {
      const text = (clipboard.readText() || '').trim();

      if (text.length !== 40) return;
      if (!/^[a-z0-9]+$/.test(text)) return;

      this.setState({ tokenInput: text });
    }

    /**
     * Handle the change event, which usually just updates the address bar's value
     *
     * @param {React.ChangeEvent<HTMLInputElement>} event
     * @memberof AddressBar
     */
    public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
      this.setState({ tokenInput: event.target.value });
    }

    get buttons() {
      const canSubmit = !!this.state.tokenInput;

      return [
        <Button
          key="done"
          disabled={!canSubmit}
          onClick={this.onSubmitToken}
          loading={this.state.verifying}
          text="Done"
          icon="log-in"
        />,
        <Button
          key="cancel"
          text="Cancel"
          icon="log-out"
          onClick={this.onClose}
        />,
      ];
    }

    get invalidWarning() {
      return (
        <>
          <Callout intent={Intent.DANGER}>
            Please provide a valid GitHub Personal Access Token
          </Callout>
          <br />
        </>
      );
    }

    public render() {
      const { isTokenDialogShowing } = this.props.appState;

      return (
        <Dialog
          isOpen={isTokenDialogShowing}
          onClose={this.onClose}
          title="GitHub Token"
        >
          <div className="bp3-dialog-body">
            <p>
              Generate a{' '}
              <a onClick={this.openGenerateTokenExternal}>
                GitHub Personal Access Token
              </a>{' '}
              and paste it here:
            </p>

            {this.state.error ? this.invalidWarning : null}

            <InputGroup
              value={this.state.tokenInput || ''}
              onFocus={this.onTokenInputFocused}
              onChange={this.handleChange}
            />
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">{this.buttons}</div>
          </div>
        </Dialog>
      );
    }
  },
);
