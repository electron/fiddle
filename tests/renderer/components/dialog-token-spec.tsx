import * as React from 'react';

import * as electron from 'electron';
import { shallow } from 'enzyme';

import { TokenDialog } from '../../../src/renderer/components/dialog-token';
import { getOctokit } from '../../../src/utils/octokit';
import { StateMock } from '../../mocks/mocks';
import { overridePlatform, resetPlatform } from '../../utils';

jest.mock('../../../src/utils/octokit');

describe('TokenDialog component', () => {
  const mockValidToken = 'testtoken1234567890123456789012345678901';
  const mockInvalidToken = 'testtoken';
  let store: StateMock;

  beforeAll(() => {
    // We render the buttons different depending on the
    // platform, so let' have a uniform platform for unit tests
    overridePlatform('darwin');
  });

  afterAll(() => {
    resetPlatform();
  });

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);
    /*
    store = {
      isTokenDialogShowing: true,
    };
    */
  });

  it('renders', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('tries to read the clipboard on focus and enters it if valid', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    (electron as any).clipboard.readText.mockReturnValueOnce(mockValidToken);
    instance.onTokenInputFocused();

    expect((electron as any).clipboard.readText).toHaveBeenCalled();
    expect(wrapper.state('tokenInput')).toBe(mockValidToken);
  });

  it('tries to read the clipboard on focus and does not enter it if too short', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    (electron as any).clipboard.readText.mockReturnValueOnce(mockInvalidToken);
    instance.onTokenInputFocused();

    expect((electron as any).clipboard.readText).toHaveBeenCalled();
    expect(wrapper.state('tokenInput')).toBe('');
  });

  it('tries to read the clipboard on focus and does not enter it if invalid', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    (electron as any).clipboard.readText.mockReturnValueOnce(
      'String with the right length not a token',
    );
    instance.onTokenInputFocused();

    expect((electron as any).clipboard.readText).toHaveBeenCalled();
    expect(wrapper.state('tokenInput')).toBe('');
  });

  it('tries to read the clipboard on focus and does not enter it if invalid', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    (electron as any).clipboard.readText.mockReturnValueOnce(undefined);
    instance.onTokenInputFocused();

    expect((electron as any).clipboard.readText).toHaveBeenCalled();
    expect(wrapper.state('tokenInput')).toBe('');
  });

  it('reset() resets the component', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    wrapper.setState({ verifying: true, tokenInput: 'hello' });
    instance.reset();

    expect(wrapper.state()).toEqual({
      verifying: false,
      error: false,
      tokenInput: '',
    });
  });

  it('onClose() resets the component', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    wrapper.setState({ verifying: true, tokenInput: 'hello' });
    instance.onClose();

    expect(wrapper.state()).toEqual({
      verifying: false,
      error: false,
      tokenInput: '',
    });
  });

  it('handleChange() handles the change event', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    wrapper.setState({ verifying: true, tokenInput: 'hello' });

    const instance: any = wrapper.instance() as any;
    instance.handleChange({ target: { value: 'hi' } } as any);

    expect(wrapper.state('tokenInput')).toBe('hi');
  });

  it('openGenerateTokenExternal() tries to open the link', () => {
    const wrapper = shallow(<TokenDialog appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    wrapper.setState({ verifying: true, tokenInput: 'hello' });
    instance.openGenerateTokenExternal();

    expect(electron.shell.openExternal).toHaveBeenCalled();
  });

  describe('onSubmitToken()', () => {
    let mockOctokit: any;
    const mockUser = {
      avatar_url: 'https://avatars.fake/hi',
      login: 'test-login',
      name: 'Test User',
    } as const;

    beforeEach(() => {
      mockOctokit = {
        authenticate: jest.fn(),
        users: {
          getAuthenticated: jest.fn().mockResolvedValue({ data: mockUser }),
        },
      };

      (getOctokit as any).mockReturnValue(mockOctokit);
    });

    it('handles missing input', async () => {
      const wrapper = shallow(<TokenDialog appState={store as any} />);
      wrapper.setState({ tokenInput: '' });
      const instance: any = wrapper.instance() as any;

      await instance.onSubmitToken();

      expect(instance.state.verifying).toBe(false);
    });

    it('tries to sign the user in', async () => {
      const wrapper = shallow(<TokenDialog appState={store as any} />);
      wrapper.setState({ tokenInput: mockValidToken });
      const instance: any = wrapper.instance() as any;

      await instance.onSubmitToken();

      expect(store.gitHubToken).toBe(mockValidToken);
      expect(store.gitHubLogin).toBe(mockUser.login);
      expect(store.gitHubName).toBe(mockUser.name);
      expect(store.gitHubAvatarUrl).toBe(mockUser.avatar_url);
    });

    it('handles an error', async () => {
      mockOctokit.users.getAuthenticated.mockReturnValue({ data: null });

      const wrapper = shallow(<TokenDialog appState={store as any} />);
      wrapper.setState({ tokenInput: mockValidToken });
      const instance: any = wrapper.instance() as any;

      await instance.onSubmitToken();

      expect(wrapper.state('error')).toBe(true);
      expect(store.gitHubToken).toEqual(null);
    });
  });
});
