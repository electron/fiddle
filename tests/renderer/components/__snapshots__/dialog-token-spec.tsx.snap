// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenDialog component renders 1`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  isOpen={false}
  onClose={[Function]}
  title="GitHub Token"
>
  <div
    className="bp3-dialog-body"
  >
    <p>
      Generate a
       
      <a
        onClick={[Function]}
      >
        GitHub Personal Access Token
      </a>
       
      and paste it here:
    </p>
    <Blueprint3.InputGroup
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        disabled={true}
        icon="log-in"
        key="done"
        loading={false}
        onClick={[Function]}
        text="Done"
      />
      <Blueprint3.Button
        icon="log-out"
        key="cancel"
        onClick={[Function]}
        text="Cancel"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;
